File: java/src/com/db4o/dg2db4o/chapter9/ObjectLock.java

package com.db4o.dg2db4o.chapter9;

import com.db4o.ObjectContainer;

public class ObjectLock {
    private String SEMAPHORE_NAME = "Semaphore: ";
    
    public boolean lock(Object obj, ObjectContainer objectContainer) {
        long id = objectContainer.ext().getID(obj);
        System.out.println("Attempting to get semaphore: " +
                SEMAPHORE_NAME + id);
        return objectContainer.ext().setSemaphore(
                SEMAPHORE_NAME + id, 1000);
    }
    
    public void release(Object obj, ObjectContainer objectContainer) {
        long id = objectContainer.ext().getID(obj);
        System.out.println("Releasing semaphore: " + SEMAPHORE_NAME + id);
        objectContainer.ext().releaseSemaphore(SEMAPHORE_NAME + id);
    }
   
}

File: csharp/Chapter09/ObjectLock.cs
using System;
using com.db4o;

namespace com.db4o.dg2db4o.chapter9
{
        public class ObjectLock
        {
            private string SEMAPHORE_NAME = "Semaphore: ";

            public Boolean Lock(Object obj, ObjectContainer objectContainer)
            {
                long id = objectContainer.Ext().GetID(obj);
                Console.WriteLine("Attempting to get semaphore: " +
                   SEMAPHORE_NAME + id);
                return objectContainer.Ext().SetSemaphore(
                    SEMAPHORE_NAME + id, 1000);
            }

            public void Release(Object obj, ObjectContainer objectContainer)
            {
                long id = objectContainer.Ext().GetID(obj);
                Console.WriteLine("Releasing semaphore: " + SEMAPHORE_NAME + id);
                objectContainer.Ext().ReleaseSemaphore(SEMAPHORE_NAME + id);
            }
        }

}
